tiledb/sm/storage_manager/consolidator.cc

```diff
@@ -374,6 +624,104 @@ void Consolidator::free_buffers(
   delete[] buffer_sizes;
 }
 
+template <class T>
+Status Consolidator::compute_next_to_consolidate(
+    const ArraySchema* array_schema,
+    const std::vector<FragmentInfo>& fragments,
+    std::vector<FragmentInfo>* to_consolidate,
+    T* union_non_empty_domains,
+    bool* coincides) const {
+  // Preparation
+  to_consolidate->clear();
+  auto min = config_.min_frags_;
+  min = (uint32_t)((min > fragments.size()) ? fragments.size() : min);
+  auto max = config_.max_frags_;
+  max = (uint32_t)((max > fragments.size()) ? fragments.size() : max);
+  auto size_ratio = config_.size_ratio_;
+
+  // Trivial case - no fragments
+  if (max == 0)
+    return Status::Ok();
+
+  // Prepare the dynamic-programming matrix. The rows are from 1 to max
+  // and the columns represent the fragments in `fragments`.
+  std::vector<std::vector<uint64_t>> m;
+  auto col_num = fragments.size();
+  auto row_num = max;
+  m.resize(row_num);
+  for (auto& row : m)
+    row.resize(col_num);
+
+  // Entry m[i][j] contains the collective size of fragments
+  // fragments[j], ..., fragments[j+i]. If the size ratio
+  // of any adjacent pair in the above list is smaller than the
+  // defined one, then the size sum of that entry is infinity (UINT64_MAX).
+  for (size_t i = 0; i < row_num; ++i) {
+    for (size_t j = 0; j < col_num; ++j) {
+      if (i == 0) {  // In the first row we store the sizes of `fragments`
+        m[i][j] = fragments[j].fragment_size_;
+      } else if (i + j >= col_num) {  // Non-valid entries
+        m[i][j] = UINT64_MAX;
+      } else {  // Every other row is computed using the previous row
+        auto ratio = (float)fragments[i + (j - 1)].fragment_size_ /
```

Comment:
Can this happen when `j == 0`? In that case we are doing `i + (0 - 1)`.

Reply:
In this `else` block, it is always `i>1`, since the case of `i==0` is handled in the first `if` block above.

Reply:
What about `j`? If it can be 0 then it will cause an OOB access into the `fragments` vector, since `size_t` is an unsigned type.

Reply:
Yeah, you are right, the parenthesis there may be problematic. I think that even with overflow this would work (as integers "wrap around"), but I will remove the parenthesis anyway.